### PR TITLE
Add append message actor

### PR DESCRIPTION
**Pull request by makadaw**

### CODE REVIEW

File: Sources/SwiftRaft/Consensus/Raft.swift [hunk 1]
```diff
@@ -225,14 +225,107 @@ extension Raft {
...
     ///
     /// 1. Reply false if term `<` currentTerm (§5.1)
     /// 2. Reply false if log doesn’t contain an entry at prevLogIndex whose term matches prevLogTerm (§5.3)
     /// 3. If an existing entry conflicts with a new one (same index but different terms), delete the existing entry and all that follow it (§5.3)
     /// 4. Append any new entries not already in the log
     /// 5. If leaderCommit > commitIndex, set commitIndex = min(leaderCommit, index of last new entry)
-    func onAppendEntries() async -> Void {
-        // TODO
+    /// Return a pair of side-effects and a response
+    func onAppendEntries<T: LogData>(_ request: AppendEntries.Request<T>) async -> ([EntriesCommand], AppendEntries.Response) {
```
Comment (ktoso) [inline Sources/SwiftRaft/Consensus/Raft.swift:292]: a bit weird return type, why not have the commands inside the response?

File: Sources/SwiftRaft/Consensus/Raft.swift [hunk 2]
```diff
@@ -225,14 +225,107 @@ extension Raft {
...
+        var commands = Array<EntriesCommand>()
+        if request.termId > self.term.id {
```
Comment (ktoso) [inline Sources/SwiftRaft/Consensus/Raft.swift:294]: a bit much nesting going on here, can we phrase this a bit nicer with `guard`?

```swift
// 1. Reply false if term `<` currentTerm (§5.1)
guard request.termID >= self.currentTermID /*self.term.id */ { 
  return self.rejectAppendEntries(....)
}
```

maybe?

File: Sources/SwiftRaft/Consensus/Raft.swift [hunk 3]
```diff
@@ -225,14 +225,107 @@ extension Raft {
...
+            // got a message with higher term, step down to a follower
+            do {
+                try term.tryToUpdateTerm(newTerm: request.termId, from: request.leaderId)
+                if !_tryMoveTo(nextState: .follower) {
+                    logger.debug("Got term greater than current and failed to move to the follower state")
```
Comment (ktoso) [inline Sources/SwiftRaft/Consensus/Raft.swift:299]: make sure to log the values, best as metadata

`logger.debug("appendEntries, incoming term >= currentTerm", metadata: ["incoming.term.id": ..., "term.id": ...])`

etc?

File: Tests/SwiftRaftTests/AppendMessageTests.swift [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        runAsyncTestAndBlock {
+            // Emulate elections
+            _ = await raft.onVoteRequest(.init(type: .vote, term: 2, candidate: 2, lastLogIndex: 0, lastLogTerm: 0))
+            let (commands, response) = await raft.onAppendEntries(self.heartBeatRequest(termId: 2, leaderId: 2))
+            XCTAssertEqual(commands.first, .resetElectionTimer)
+            XCTAssertEqual(response.termId, 2)
+            XCTAssertEqual(response.success, true)
+        }
+    }
+}
```
Comment (ktoso) [inline Tests/SwiftRaftTests/AppendMessageTests.swift:37]: cool yeah, this way we can test them nicely in isolation 👍

**Review by ktoso — APPROVED** [review]
I have not reviewed protocol details, a bit though on time now but the shape is looking okey :-)

File: Sources/SwiftRaft/Consensus/Raft.swift (hunk 1 above)
Comment (makadaw) [reply Sources/SwiftRaft/Consensus/Raft.swift:292]: Good idea. My first approach was to have a command `response`, but in this case Response become kind of optional, but inverse it make a lot of sens

File: Sources/SwiftRaft/Consensus/Raft.swift (hunk 3 above)
Comment (makadaw) [reply Sources/SwiftRaft/Consensus/Raft.swift:299]: Wil add

**Comment by makadaw** [comment]
Created new MR into main https://github.com/makadaw/swift-raft/pull/8

**Pull request closed without merging**